### DISCUSSION TITLE
"JavaScript heap out of memory" when running tsc after migrating to the latest versions of i18next and react-i18next

### DISCUSSION THREAD

**Discussion by szimek** [i18next/i18next#2246] [Q&A]
Hi!

I've got a project that was using `i18next` 21.x and `react-i18next` 11.x. I migrated everything to the latest versions (23.x and 15.x respectively) and it all works. However, I noticed that I wasn't getting TS errors when I provided a non-existing key to the translation function.

After updating the `i18n.d.ts` file (basically replacing `react-i18next` with `i18next`) I'm getting TS errors again in my editor, however running `tsc` fails now with "JavaScript heap out of memory" error.

My `i18n.d.ts` file looks like this:

```ts
import 'i18next'

import type resources from '@/locales/en'
import type { defaultNS } from './i18n'

declare module 'i18next' {
  interface CustomTypeOptions {
    defaultNS: typeof defaultNS
    resources: typeof resources
  }
}
```

We've got almost 4000 keys spread across 40 `.json` files (that's what `resources` is in the code above) and we'll most likely double it soon. We're using TS 5.6.2 and Node 20.12.0. Besides increasing the memory limit for Node (`tsc` works with `NODE_OPTIONS=--max_old_space_size=8192`), is there any other way to fix it, especially that it used to work with older versions of both libraries?

One more thing I've just noticed. We've got a similar code in a few places (unit tests):
```ts
  const tCommon = ((key: string) => `localized ${key}`) as TFunction<'common', undefined>
```
that is causing `Type instantiation is excessively deep and possibly infinite.` error after the upgrade.

**Comment by ahrjarrett** [comment]
This issue should be fixed in [v25.4.0](https://github.com/i18next/i18next/releases/tag/v25.4.0) with type option `enableSelector` set to `"optimize"`
